README.md
# sesdev - deploy and manage SES/Ceph clusters<br/>

`sesdev` is a CLI tool to deploy Ceph clusters (both the upstream and SUSE
downstream versions).

This tool uses [Vagrant](https://www.vagrantup.com/) behind the scenes to create
the VMs and run the deployment scripts.

## Build Status

### Travis

[![Travis Build Status](https://travis-ci.org/SUSE/sesdev.svg?branch=master)](https://travis-ci.org/SUSE/sesdev)

The Travis CI tests that the Python source code of `sesdev` compiles and has no
linter issues.

### Jenkins

[![Jenkins Build Status](https://ceph-ci.suse.de/job/sesdev-integration/job/master/badge/icon)](https://ceph-ci.suse.de/job/sesdev-integration/job/master/)

The Jenkins CI tests that `sesdev` can be used to deploy a single-node Ceph
15.2.x ("Octopus") cluster in an openSUSE Leap 15.2 environment.

## Table of Contents

[//]: # (To generate a new TOC, first install https://github.com/ekalinin/github-markdown-toc)
[//]: # (and then run "gh-md-toc README.md")
[//]: # (the new TOC will appear on stdout: the expectation is that the maintainer will do the rest.)

* [Installation](#installation)
   * [Install sesdev on openSUSE or SUSE Linux Enterprise](#install-sesdev-on-opensuse-or-suse-linux-enterprise)
      * [Install KVM/QEMU and Libvirt](#install-kvmqemu-and-libvirt)
      * [Add user to libvirt group](#add-user-to-libvirt-group)
      * [Install Vagrant](#install-vagrant)
         * [Install Vagrant on openSUSE Leap 15.2, SLE-15-SP2 or Tumbleweed](#install-vagrant-on-opensuse-leap-152-sle-15-sp2-or-tumbleweed)
         * [Install Vagrant on openSUSE Leap 15.1, SLE-15-SP1](#install-vagrant-on-opensuse-leap-151-sle-15-sp1)
         * [Vagrant RPM from Hashicorp website](#vagrant-rpm-from-hashicorp-website)
      * [Install sesdev from package](#install-sesdev-from-package)
   * [Install sesdev on Fedora Linux](#install-sesdev-on-fedora-linux)
      * [Install KVM/QEMU and Libvirt](#install-kvmqemu-and-libvirt-1)
      * [Install sesdev from package](#install-sesdev-from-package-1)
   * [Install sesdev on Debian/Ubuntu](#install-sesdev-on-debianubuntu)
   * [Install sesdev from source](#install-sesdev-from-source)
      * [Running the unit tests](#running-the-unit-tests)
* [Usage](#usage)
   * [Create/deploy a Ceph cluster](#createdeploy-a-ceph-cluster)
      * [Bare bone cluster](#bare-bone-cluster)
      * [CaaSP (with or without Rook/Ceph/SES)](#caasp-with-or-without-rookcephses)
         * [CaaSP k8s cluster](#caasp-k8s-cluster)
         * [CaaSP with Rook/Ceph/SES](#caasp-with-rookcephses)
         * [CaaSP on just one node](#caasp-on-just-one-node)
      * [On a remote libvirt server via SSH](#on-a-remote-libvirt-server-via-ssh)
      * [Using salt instead of DeepSea/ceph-salt CLI](#using-salt-instead-of-deepseaceph-salt-cli)
      * [With a FQDN environment](#with-a-fqdn-environment)
      * [Without the devel repo](#without-the-devel-repo)
      * [With an additional custom zypper repo](#with-an-additional-custom-zypper-repo)
      * [With a set of custom zypper repos completely replacing the default repos](#with-a-set-of-custom-zypper-repos-completely-replacing-the-default-repos)
      * [With custom image paths](#with-custom-image-paths)
      * [With custom default roles](#with-custom-default-roles)
      * [config.yaml examples](#configyaml-examples)
         * [octopus from filesystems:ceph:octopus](#octopus-from-filesystemscephoctopus)
         * [octopus from filesystems:ceph:octopus&#8203;:upstream](#octopus-from-filesystemscephoctopusupstream)
         * [ses7 from Devel:Storage:7.0](#ses7-from-develstorage70)
         * [ses7 from Devel:Storage:7.0:CR](#ses7-from-develstorage70cr)
      * [With wire encryption](#with-wire-encryption)
      * [Deploying non-SUSE environments](#deploying-non-suse-environments)
         * [Ubuntu "Bionic Beaver" 18.04](#ubuntu-bionic-beaver-1804)
      * [Introspect existing deployments](#introspect-existing-deployments)
         * [List all existing deployments and their overall status](#list-all-existing-deployments-and-their-overall-status)
         * [Get status of individual nodes in an existing deployment](#get-status-of-individual-nodes-in-an-existing-deployment)
         * [Show details of an existing deployment](#show-details-of-an-existing-deployment)
         * [Show roles of nodes in an existing deployment](#show-roles-of-nodes-in-an-existing-deployment)
   * [List existing deployments](#list-existing-deployments)
   * [SSH access to a cluster](#ssh-access-to-a-cluster)
   * [Copy files into and out of a cluster](#copy-files-into-and-out-of-a-cluster)
   * [Services port-forwarding](#services-port-forwarding)
   * [Replace ceph-salt](#replace-ceph-salt)
   * [Replace MGR modules](#replacing-mgr-modules)
   * [Add a repo to a cluster](#add-a-repo-to-a-cluster)
   * [Link two clusters together](#link-two-clusters-together)
   * [Temporarily stop a cluster](#temporarily-stop-a-cluster)
   * [Destroy a cluster](#destroy-a-cluster)
   * [Run "make check"](#run-make-check)
      * [Run "make check" on Tumbleweed from upstream "master" branch](#run-make-check-on-tumbleweed-from-upstream-master-branch)
      * [Run "make check" on openSUSE Leap 15.2 from upstream "octopus" branch](#run-make-check-on-opensuse-leap-152-from-upstream-octopus-branch)
      * [Run "make check" on SLE-15-SP2 from downstream "ses7" branch](#run-make-check-on-sle-15-sp2-from-downstream-ses7-branch)
      * [Other "make check" scenarios](#other-make-check-scenarios)
* [Common pitfalls](#common-pitfalls)
   * [Domain about to create is already taken](#domain-about-to-create-is-already-taken)
   * [Storage pool not found: no storage pool with matching name 'default'](#storage-pool-not-found-no-storage-pool-with-matching-name-default)
   * [When sesdev deployments get destroyed, virtual networks get left behind](#when-sesdev-deployments-get-destroyed-virtual-networks-get-left-behind)
   * [sesdev destroy reported an error](#sesdev-destroy-reported-an-error)
   * ["Failed to connect socket" error when attempting to use remote libvirt server](#failed-to-connect-socket-error-when-attempting-to-use-remote-libvirt-server)
   * [mount.nfs: Unknown error 521](#mountnfs-unknown-error-521)
   * [Problems accessing dashboard on remote sesdev](#problems-accessing-dashboard-on-remote-sesdev)
   * [Error creating IPv6 cluster](#error-creating-ipv6-cluster)
   * [Failed to initialize libnetcontrol](#failed-to-initialize-libnetcontrol)
* [Contributing](#contributing)


## Installation

First, you should have both [QEMU](https://www.qemu.org/) and
[Libvirt](https://libvirt.org/) installed in some machine to host the VMs
created by sesdev (using Vagrant behind the scenes).

Installable packages for various Linux distributions like Fedora or openSUSE can
be found on the [openSUSE Build Service](https://software.opensuse.org//download.html?project=filesystems%3Aceph&package=sesdev)
(OBS).

### Install sesdev on openSUSE or SUSE Linux Enterprise

#### Install KVM/QEMU and Libvirt

First, if on SUSE Linux Enterprise, make sure you have the Server Applications
Module available on the system. (Links to internal repos are available
[here](https://confluence.suse.com/display/applications/Vagrant+and+Vagrant+Boxes+for+openSUSE+and+SLE#VagrantandVagrantBoxesforopenSUSEandSLE-openSUSELeap15.2andSLE15SP2).)

Run the following commands as root:

```
# zypper -n install -t pattern kvm_server kvm_tools
# systemctl enable libvirtd
# systemctl restart libvirtd
```

#### Add user to libvirt group

If you are running libvirt on the same machine where you installed sesdev, add
your user to the "libvirt" group to avoid "no polkit agent available" errors
when vagrant attempts to connect to the libvirt daemon:

```
# groupadd libvirt
groupadd: group 'libvirt' already exists
# usermod -a -G libvirt $USER
```

Log out, and then log back in. You should now be a member of the "libvirt"
group.

#### Install Vagrant

sesdev needs Vagrant to work. Vagrant can be installed in a number of ways,
depending on your environment:

##### Install Vagrant on openSUSE Leap 15.2, SLE-15-SP2 or Tumbleweed

On very new OSes like these, Vagrant is included in the operating system's base
repos. Just install the ``vagrant`` and ``vagrant-libvirt`` packages.

For SLE-15-SP2, the packages are available via the
[SUSE Package Hub](https://packagehub.suse.com/). (Links to internal repos are available
[here](https://confluence.suse.com/display/applications/Vagrant+and+Vagrant+Boxes+for+openSUSE+and+SLE#VagrantandVagrantBoxesforopenSUSEandSLE-openSUSELeap15.2andSLE15SP2).)

##### Install Vagrant on openSUSE Leap 15.1, SLE-15-SP1

To install Vagrant on these systems, run the following commands as root:

```
# zypper ar https://download.opensuse.org/repositories/Virtualization:/vagrant/<repo> vagrant_repo
# zypper ref
# zypper -n install vagrant vagrant-libvirt
```

Where `<repo>` can be any of the openSUSE build targets currently enabled for
the [Virtualization:vagrant/vagrant package in the openSUSE Build Service](https://build.opensuse.org/package/show/Virtualization:vagrant/vagrant).

Be aware that ``Virtualization:vagrant`` is a development project where updates
to the latest official openSUSE vagrant packages are prepared. That means the
vagrant packages in this repo will tend to be new and, sometimes, even broken.
In that case, read on to the next section.

##### Vagrant RPM from Hashicorp website

If you find that, for whatever reason, you cannot get a working vagrant package
from OBS, it is possible to install vagrant from the official RPMs published on
[the Hashicorp website](https://www.vagrantup.com/downloads.html).

To install vagrant and its libvirt plugin from Hashicorp, the following
procedure has been known to work (run the commands as root):

1. download vagrant RPM from https://releases.hashicorp.com/vagrant/
2. install make (``zypper install make``)
3. install vagrant (``rpm -i <the RPM you just downloaded>``)
4. delete file that causes libvirt plugin compilation to fail
   (``rm /opt/vagrant/embedded/lib/libreadline.so.7``)

Finally, run the following command as the user you run sesdev with:

    vagrant plugin install vagrant-libvirt

#### Install sesdev from package

sesdev itself can be installed either from package or from source. If you
prefer to install from package, follow the instructions in this section. If you
prefer to install from source, skip down to the "Install sesdev from source"
section.

Run the following commands as root:

```
# zypper ar https://download.opensuse.org/repositories/filesystems:/ceph/<repo> filesystems_ceph
# zypper ref
# zypper install sesdev
```

Where `<repo>` can be any of the openSUSE build targets currently enabled for
the [sesdev package in the openSUSE Build Service](https://build.opensuse.org/package/show/filesystems:ceph/sesdev).

At this point, sesdev should be installed and ready to use: refer to the
[Usage](#usage) chapter, below, for further information.

### Install sesdev on Fedora Linux

#### Install KVM/QEMU and Libvirt

Run the following commands as root:

```
# dnf install qemu-common qemu-kvm libvirt-daemon-kvm \
libvirt-daemon libvirt-daemon-driver-qemu vagrant-libvirt
# systemctl enable libvirtd
# systemctl restart libvirtd
```

#### Install sesdev from package

Run the following commands as root:

```
# dnf config-manager --add-repo \
https://download.opensuse.org/repositories/filesystems:/ceph/<distro>/filesystems:ceph.repo
# dnf install sesdev
```

Where `<distro>` can be any of the Fedora build targets currently enabled for
the [sesdev package in the openSUSE Build Service](https://build.opensuse.org/package/show/filesystems:ceph/sesdev).

At this point, sesdev should be installed and ready to use: refer to the
[Usage](#usage) chapter, below, for further information.

### Install sesdev on Debian/Ubuntu

sesdev is known to work on recent Ubuntu versions, but there is no package
for it: you have to install from source. Follow the instructions given in
[Install sesdev from source](#install-sesdev-from-source).

### Install sesdev from source

sesdev itself can be installed either from package or from source. If you
prefer to install from source, follow the instructions in this section. If you
prefer to install from package, scroll up to the "Install sesdev from package"
section for your operating system.

sesdev uses the libvirt API Python bindings, and these cannot be installed via
pip unless the RPM packages "gcc", "python3-devel", and "libvirt-devel" are
installed, first. Also, in order to clone the sesdev git repo, the "git-core"
package is needed. So, before proceeding, make sure that all of these packages
are installed in the system:

#### openSUSE

```
# zypper -n install gcc git-core libvirt-devel python3-devel python3-virtualenv
```

#### Debian / Ubuntu

```
# apt-get install -y git gcc libvirt-dev \
virtualenv python3-dev python3-venv python3-virtualenv
```

#### Fedora

```
# dnf install -y git-core gcc libvirt-devel \
python3-devel python3-virtualenv
```

Now you can proceed to clone the sesdev source code repo and bootstrap it:

```
$ git clone https://github.com/SUSE/sesdev.git
$ cd sesdev
$ ./bootstrap.sh
```

Before you can use `sesdev`, you need to activate the Python virtual environment
created by the `bootstrap.sh` script. The script tells you how to do this, but
we'll give the command here, anyway:

```
source venv/bin/activate
```

At this point, sesdev should be installed and ready to use: refer to the
[Usage](#usage) chapter, below, for further information.

To leave the virtual environment, simply run:

```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project aspires to adhere to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.12.1] - 2021-04-26

### Added
- contrib/nukevolz.sh: primitive script to recover wasted disk space (PR #593)

### Fixed
- zypper: remove python3-base if it's installed (PR #594)
- templates: disallow global ID reclaim (PR #596)

### Removed
- deployment: Remove redundant node network code (PR #592)

## [1.12.0] - 2021-03-22

### Added
- provision: enable coredumps (PR #566)
- sesdev: support custom repo URLs pointing to .repo file (PR #571)
- cephadm: add 'period update' in radosgw deployment (PR #579)
- tox: add support for Python 3.8 (PR #582)

### Fixed
- sesdev.spec: fix missing qa/common/rgw.sh (PR #572)
- contrib/upgrade-demo-ses6-to-ses7.sh: fix (PR #573)
- zypper: be more careful when installing systemd-coredump (PR #577)
- sesdev: replace fancy list comprehension with simpleton code (PR #581)
- deployment: saner processing of --deepsea-repo, --deepsea-branch (PR #584)
- zypper.j2: remove Python 2 even on SLE-12-SP3 (PR #585)

### Changed
- provision.sh: persist systemd journal in SES7 only (PR #569)
- ceph-salt: no cephadm for nodes with no sesdev roles (PR #570)
- constant.py: drop nfs from octopus default roles (PR #587)
- nautilus_pre_stage_4.sh.j2: adapt to syntax change in Nautilus 14.2.17
  (PR #589)
- qa/common/json.sh: adapt to change in ceph status json format (PR #590)

## [1.11.0] - 2020-12-03

### Added
- Implement --rgw-ssl option for deploying RGWs with SSL enabled (PR #557)

### Fixed
- zypper: add repos when needed (PR #558)
- ses7: use product container image when --product given (PR #561)
- Handle image_paths_product, image_paths_devel overrides in config.yaml (PR #564)

### Changed
- seslib/constant.py: add SES7 Update repo (PR #565)

## [1.10.0] - 2020-11-13

### Added
- contrib/standalone.sh: add --fqdn test cases (PR #523)
- Implement --no-apparmor option (PR #543)
- Expose Ceph Dashboard on CaaSP SES deployment (PR #553)
- doc/README.md: add common pitfall "Failed to intialize libnetcontrol"
  (PR #554) 

### Changed
- spec: pin sesdev-qa version to sesdev version (PR #551)
- constant,zypper: use download.nue.suse.com instead of download.suse.de
  (PR #552)
- deployment_day_2.sh: create CephFS using documented procedure (PR #555)

## [1.9.1] - 2020-10-23

### Fixed
- sesdev --devel flag: let click deduct type (PR #550)

## [1.9.0] - 2020-10-23

### Added
- constant: add missing SLE-15-SP[12] Product repos (PR #530)
- Implement "sesdev upgrade" subcommand (PR #531)
- Implement "sesdev reboot DEP_ID NODE" command (PR #532)
- contrib/upgrade-demo-ses6-to-ses7.sh: add upgrade demo script (PR #535)
- provision.sh: add helper_scripts Jinja template (PR #540)

### Fixed
- exceptions: have SesDevException return optional exit code (PR #538)
- deployment: wait for rebooted node to complete boot sequence (PR #539)
- sesdev: fix --filestore regression (PR #545)
- cephadm/deployment_day_2.sh.j2: do not fail on defective --dry-run (PR #546)
- setup.cfg: include seslib/templates/cephadm/ directory in packaging (PR #548)

### Changed
- templates/zypper: do not install supportutils-plugin-ses in SES{5,6,7} (PR #526)

### Removed
- deepsea/nautilus_pre_stage_0.sh: refrain from patching DeepSea (PR #547)

## [1.8.0] - 2020-10-08

### Added
- Add support for IPv6 (PR #56)
- contrib/deepsea_drive_replace.sh: new test script (PR #306)
- ceph_salt: deploy and smoke-test Prometheus (PR #418)
- ceph-salt: deploy and smoke-test Grafana (PR #475) 
- ceph-salt: deploy and smoke-test alertmanager (PR #480)
- ceph-salt: deploy and smoke-test node-exporter (PR #496)
- Support official Vagrant boxes from Vagrant Cloud (PR #476)
- new msgr2 secure options (PR #481)
- Add ability to deploy Ceph on Ubuntu "Bionic Beaver" 18.04 (PR #486)
- add label for ses deployment on caasp (PR #488)
- Implement --no-provision option to deploy "bare" VMs (PR #490)
- Implement "sesdev link" for linking two clusters (PR #498)
- Implement "sesdev status" subcommand (PR #502)
- Implement "sesdev show DEP_ID --nodes-with-role=ROLE" (PR #506)
- Make "sesdev box list" and "sesdev box remove" accept globs (PR #517)
- create: implement --fqdn option, specify FQDN when appropriate (PR #519, PR #522)

### Removed
- replace --deploy/--no-deploy option with --provision/--no-provision (PR #490)
- caasp4: remove anti-affinity for mons for single node (PR #495) 
- Revert "ceph_salt_deployment: add 1-minute grace period" (PR #521)

### Fixed
- ceph-salt: set some network-related config params explicitly (PR #482)
- deployment: do not assert if starting an existing deployment (PR #493)
- deployment: do not populate public_network with "0/24" (PR #512)
- qa: wait for grafanas to show up before counting them (PR #514)
- qa: wait longer for cluster to become healthy (PR #515)
- Add makecheck-specific repos only in "sesdev create makecheck" (PR #520)

### Changed
- zypper: install rbd-nbd on all Leap 15.2/SLE-15-SP2 nodes (PR #484)
- Update pacific repos (PR #489)
- qa/nfs: do not attempt to mount NFS export on "pacific" (PR #492)
- deployment: use "vagrant up --provision" to start cluster (PR #501)
- cleanup: move cephadm Day 2 provisioner to its own subdirectory (PR #513) 
- tox: run pip with --use-feature=2020-resolver (PR #516)

## [1.7.0] - 2020-09-13

### Added
- ceph_salt_deployment: do "ceph orch apply --dry-run" (PR #406)
- single node CaaSP cluster (PR #427)
- If --ssd option is given, attempt to make first additional disk
  non-rotational (PR #448)
- Implement "sesdev box remove --all" feature (PR #451)
- qa: support openSUSE Tumbleweed (PR #457)

### Removed
- Drop ceph-salt '/system_update' config (PR #447)

### Fixed
- replace-mgr-modules: refrain from introducing Python 2 (PR #431)

sesdev.py
import logging
import sys

import click
import pkg_resources
import seslib
from seslib.exceptions import SesDevException


logger = logging.getLogger(__name__)


def sesdev_main():
    try:
        # pylint: disable=unexpected-keyword-arg
        cli(prog_name='sesdev')
    except SesDevException as ex:
        logger.exception(ex)
        click.echo(str(ex))


def _decorator_composer(decorator_list, func):
    if not decorator_list:
        return func
    head, *tail = decorator_list
    return _decorator_composer(tail, head(func))


def libvirt_options(func):
    click_options = [
        click.option('--libvirt-host', type=str, default=None,
                     help='Hostname of the libvirt machine'),
        click.option('--libvirt-user', type=str, default=None,
                     help='Username for connecting to the libvirt machine'),
        click.option('--libvirt-storage-pool', type=str, default=None, help='Libvirt storage pool'),
    ]
    return _decorator_composer(click_options, func)


def deepsea_options(func):
    click_options = [
        click.option('--deepsea-cli/--salt-run', default=True,
                     help="Use deepsea-cli or salt-run to execute DeepSea stages"),
        click.option('--stop-before-deepsea-stage', type=int, default=None,
                     help='Allows to stop deployment before running the specified DeepSea stage'),
        click.option('--deepsea-repo', type=str, default=None, help='DeepSea Git repo URL'),
        click.option('--deepsea-branch', type=str, default=None, help='DeepSea Git branch'),
    ]
    return _decorator_composer(click_options, func)


def common_create_options(func):
    click_options = [
        click.option('--roles', type=str, default=None,
                     help='List of roles for each node. Example for two nodes: '
                          '[admin, client, prometheus],[storage, mon, mgr]'),
        click.option('--os', type=click.Choice(['leap-15.1', 'leap-15.2', 'tumbleweed',
                                                'sles-12-sp3', 'sles-15-sp1', 'sles-15-sp2']),
                     default=None, help='OS (open)SUSE distro'),
        click.option('--vagrant-box', type=str, default=None,
                     help='Vagrant box to use in deployment'),
        click.option('--deploy/--no-deploy', default=True,
                     help="Don't run the deployment phase. Just generated the Vagrantfile"),
        click.option('--cpus', default=None, type=int,
                     help='Number of virtual CPUs for the VMs'),
        click.option('--ram', default=None, type=int,
                     help='Amount of RAM for each VM in gigabytes'),
        click.option('--disk-size', default=None, type=int,
                     help='Size in gigabytes of storage disks (used by OSDs)'),
        click.option('--num-disks', default=None, type=int,
                     help='Number of storage disks in OSD nodes'),
        click.option('--single-node/--no-single-node', default=False,
                     help='Deploy a single node cluster. Overrides --roles'),
        click.option('--repo', multiple=True, type=str, default=None,
                     help='Zypper repo URL. The repo will be added to each node.')
    ]
    return _decorator_composer(click_options, func)


def _parse_roles(roles):
    roles = [r.strip() for r in roles.split(",")]
    _roles = []
    _node = None
    for r in roles:
        r = r.strip()
        if r.startswith('['):
            _node = []
            if r.endswith(']'):
                r = r[:-1]
                _node.append(r[1:])
                _roles.append(_node)
            else:
                _node.append(r[1:])
        elif r.endswith(']'):
            _node.append(r[:-1])
            _roles.append(_node)
        else:
            _node.append(r)
    return _roles


def _print_log(output):
    sys.stdout.write(output)
    sys.stdout.flush()


def _silent_log(output):
    pass


def _abort_if_false(ctx, param, value):
    if not value:
        ctx.abort()


@click.group()
@click.option('-w', '--work-path', required=False,
              type=click.Path(exists=True, dir_okay=True, file_okay=False),
              help='Filesystem path to store deployments')
@click.option('-c', '--config-file', required=False,
              type=click.Path(exists=True, dir_okay=False, file_okay=True),
              help='Configuration file location')
@click.option('--debug/--no-debug', default=False)
@click.option('--log-file', type=str, default=None)
@click.version_option(pkg_resources.get_distribution('sesdev'), message="%(version)s")
def cli(work_path=None, config_file=None, debug=False, log_file=None):
    """
    Welcome to the sesdev tool.

    Usage example:

    # Deployment of single node SES6 cluster:

        $ sesdev create ses6 --single-node my_ses6_cluster

    # Deployment of Octopus cluster with deepsea where each storage node contains 4 10G disks for
OSDs:

        \b
$ sesdev create octopus --roles="[admin, mon, mgr], \\
       [storage, mon, mgr, mds], [storage, mon, mds]" \\
       --use-deepsea --num-disks=4 --disk-size=10 my_octopus_cluster

    """
    if log_file:
        logging.basicConfig(format='%(asctime)s [%(levelname)s] [%(name)s] %(message)s',
                            filename=log_file, filemode='w',
                            level=logging.INFO if not debug else logging.DEBUG)
    else:
        logging.basicConfig(format='%(asctime)s [%(levelname)s] [%(name)s] %(message)s',
                            level=logging.CRITICAL)

    if work_path:
        logger.info("Working path: %s", work_path)
        seslib.GlobalSettings.WORKING_DIR = work_path

    if config_file:
        logger.info("Config file: %s", config_file)
        seslib.GlobalSettings.CONFIG_FILE = config_file


@cli.command()
def list():
    """
    Lists all the available deployments.
    """
    deps = seslib.Deployment.list(True)
    click.echo("| {:^11} | {:^15} | {:^60} |".format("Deployments", "Status", "VMs"))
    click.echo("{}".format('-' * 96))

    def _status(nodes):
        s = nodes['admin'].status
        for n in nodes.values():
            if n.status == 'running' and s == 'not deployed':
                s = 'partially deployed'
            elif n.status == 'stopped' and s == 'running':
                s = 'partially running'
            elif n.status == 'suspended' and s == 'running':
                s = 'partially running'
            elif n.status == 'running' and s == 'stopped':
                s = 'partially running'
            elif n.status == 'running' and s == 'suspended':
                s = 'partially running'
        return s

    for dep in deps:
        st = _status(dep.nodes)
        click.echo("| {:<11} | {:<15} | {:<60} |".format(dep.dep_id, st, ", ".join(dep.nodes)))
    click.echo()


@cli.group()
def create():
    """
    Creates a new Vagrant based SES cluster.

    It creates a deployment directory in <working_directory>/<deployment_id>
    with a Vagrantfile inside, and calls `vagrant up` to start the deployment.

    By default <working_directory> is located in `~/.sesdev`.

    Checks all the options available with:

    $ sesdev create --help
    """
    pass


def _gen_settings_dict(version, roles, os, num_disks, single_node, libvirt_host, libvirt_user,
                       libvirt_storage_pool, deepsea_cli, stop_before_deepsea_stage, deepsea_repo,
                       deepsea_branch, repo, cpus, ram, disk_size, vagrant_box):

    settings_dict = {}
    if not single_node and roles:
        settings_dict['roles'] = _parse_roles(roles)
    elif single_node:
        settings_dict['roles'] = [["admin", "storage", "mon", "mgr", "prometheus", "grafana", "mds",
                                  "igw", "rgw", "ganesha"]]

    if os:
        settings_dict['os'] = os

    if num_disks:
        if single_node and num_disks < 3:
            num_disks = 3
        settings_dict['num_disks'] = num_disks
    elif single_node:
        settings_dict['num_disks'] = 3

    if cpus:
        settings_dict['cpus'] = cpus

    if ram:
        settings_dict['ram'] = ram

    if disk_size:
        settings_dict['disk_size'] = disk_size

    if libvirt_host:
        settings_dict['libvirt_host'] = libvirt_host

    if libvirt_user:
        settings_dict['libvirt_user'] = libvirt_user

    if libvirt_storage_pool:
        settings_dict['libvirt_storage_pool'] = libvirt_storage_pool

    if deepsea_cli is not None:
        settings_dict['use_deepsea_cli'] = deepsea_cli

    if stop_before_deepsea_stage is not None:
        settings_dict['stop_before_stage'] = stop_before_deepsea_stage

    if deepsea_repo:
        settings_dict['deepsea_git_repo'] = deepsea_repo

    if deepsea_branch:
        settings_dict['deepsea_git_branch'] = deepsea_branch

    if version is not None:
        settings_dict['version'] = version

    if repo:
        settings_dict['repos'] = [r for r in repo]

    if vagrant_box:
        settings_dict['vagrant_box'] = vagrant_box

    return settings_dict


def _create_command(deployment_id, deploy, settings_dict):
    settings = seslib.Settings(**settings_dict)
    dep = seslib.Deployment.create(deployment_id, settings)
    click.echo("=== Creating deployment with the following configuration ===")
    click.echo(dep.status())
    if deploy:
        if click.confirm('Do you want to continue with the deployment?', default=True):
            dep.start(_print_log)
            click.echo("=== Deployment Finished ===")
            click.echo()
            click.echo("You can login into the cluster with:")
            click.echo()
            click.echo("  $ sesdev ssh {}".format(deployment_id))
            click.echo()
            if dep.settings.version == 'ses5':
                click.echo("Or, access openATTIC with:")
                click.echo()
                click.echo("  $ sesdev tunnel {} openattic".format(deployment_id))
            else:
                click.echo("Or, access the Ceph Dashboard with:")
                click.echo()
                click.echo("  $ sesdev tunnel {} dashboard".format(deployment_id))
                click.echo()
        else:
            dep.destroy(_silent_log)


@create.command()
@click.argument('deployment_id')
@common_create_options
@deepsea_options
@libvirt_options
def ses5(deployment_id, deploy, **kwargs):
    """
    Creates a SES5 cluster using SLES-12-SP3
    """
    settings_dict = _gen_settings_dict('ses5', **kwargs)
    if 'roles' in settings_dict:
        settings_dict['roles'][0].append("openattic")
    _create_command(deployment_id, deploy, settings_dict)


@create.command()
@click.argument('deployment_id')
@common_create_options
@deepsea_options
@libvirt_options
def ses6(deployment_id, deploy, **kwargs):
    """
    Creates a SES6 cluster using SLES-15-SP1
    """
    settings_dict = _gen_settings_dict('ses6', **kwargs)
    _create_command(deployment_id, deploy, settings_dict)


@create.command()
@click.argument('deployment_id')
@common_create_options
@deepsea_options
@libvirt_options
@click.option("--use-deepsea/--use-orchestrator", default=False,
              help="Use deepsea to deploy SES7 instead of SSH orchestrator")
def ses7(deployment_id, deploy, use_deepsea, **kwargs):
    """
    Creates a SES7 cluster using SLES-15-SP2
    """
    settings_dict = _gen_settings_dict('ses7', **kwargs)
    if use_deepsea:
        settings_dict['deployment_tool'] = 'deepsea'
    _create_command(deployment_id, deploy, settings_dict)


@create.command()
@click.argument('deployment_id')
@common_create_options
@deepsea_options
@libvirt_options
def nautilus(deployment_id, deploy, **kwargs):
    """
    Creates a Ceph Nautilus cluster using openSUSE Leap 15.1 and packages
    from filesystems:ceph:nautilus OBS project.
    """
    settings_dict = _gen_settings_dict('nautilus', **kwargs)
    _create_command(deployment_id, deploy, settings_dict)


@create.command()
@click.argument('deployment_id')
@common_create_options
@deepsea_options
@libvirt_options
@click.option("--use-deepsea/--use-orchestrator", default=False,
              help="Use deepsea to deploy Ceph Octopus instead of SSH orchestrator")
def octopus(deployment_id, deploy, use_deepsea, **kwargs):
    """
    Creates a Ceph Octopus cluster using openSUSE Leap 15.2 and packages
    from filesystems:ceph:octopus OBS project.
    """
    settings_dict = _gen_settings_dict('octopus', **kwargs)
    settings_dict['deployment_tool'] = 'deepsea'
    _create_command(deployment_id, deploy, settings_dict)


@cli.command()
@click.argument('deployment_id')
@click.option('--force', is_flag=True, callback=_abort_if_false, expose_value=False,
              help='Allow to destroy the deployment without user confirmation',
              prompt='Are you sure you want to destroy the cluster?')
def destroy(deployment_id):
    """
    Destroys the deployment named DEPLOYMENT_ID by destroying the VMs and deletes the
    deployment directory.
    """
    dep = seslib.Deployment.load(deployment_id)
    dep.destroy(_print_log)


@cli.command()
@click.argument('deployment_id')
@click.argument('node', required=False)
def ssh(deployment_id, node=None):
    """
    Opens an SSH shell to node NODE in deployment DEPLOYMENT_ID.
    If the node is not specified, an SSH shell is opened on the "admin" node.
    """
    dep = seslib.Deployment.load(deployment_id)
    dep.ssh(node if node else 'admin')


@cli.command()
@click.argument('deployment_id')
@click.argument('node', required=False)
def stop(deployment_id, node=None):
    """
    Stops the VMs of the deployment DEPLOYMENT_ID.
    """
    dep = seslib.Deployment.load(deployment_id)
    dep.stop(_print_log, node)


@cli.command()
@click.argument('deployment_id')
@click.argument('node', required=False)
def start(deployment_id, node=None):
    """
    Stars the VMs of the deployment DEPLOYMENT_ID.

    If cluster was not yet deployed (if was created with the --no-deploy flag), it will
    start the deployment of the cluster.
    """
    dep = seslib.Deployment.load(deployment_id)
    dep.start(_print_log, node)


@cli.command()
@click.argument('deployment_id')
def show(deployment_id):
    """
    Shows the information of deployment DEPLOYMENT_ID
    """
    dep = seslib.Deployment.load(deployment_id)
    click.echo(dep.status())


@cli.command()
@click.argument('deployment_id')
@click.option('--force', is_flag=True, callback=_abort_if_false, expose_value=False,
              help='Allow to redeploy the cluster without user confirmation',
              prompt='Are you sure you want to redeploy the cluster?')
def redeploy(deployment_id):
    """
    Destroys the VMs of the deployment DEPLOYMENT_ID and deploys again the cluster
    from scratch with the same configuration.
    """
    dep = seslib.Deployment.load(deployment_id)
    dep.destroy(_print_log)
    dep = seslib.Deployment.create(deployment_id, dep.settings)
    dep.start(_print_log)


@cli.command()
@click.argument('deployment_id')
@click.argument('service', type=click.Choice(['dashboard', 'grafana', 'openattic']), required=False)
@click.option('--node', default='admin', type=str, show_default=True,
              help='The node where we want to create the tunnel to')
@click.option('--remote-port', default=None, type=int,
              help='The service port in the remote machine')
@click.option('--local-port', default=None, type=int, help='The local port for the tunnel')
@click.option('--local-address', default='localhost', type=str, show_default=True,
              help='The local address to bind the tunnel')
def tunnel(deployment_id, service=None,  node=None, remote_port=None, local_port=None,
           local_address=None):
    """
    Creates an SSH port forwarding for the services that are running in the
    deployment DEPLOYMENT_ID.

    If SERVICE is not specified, you can use the --remote-port and --node to forward
    a generic service.
    """
    if service:
        click.echo("Opening tunnel to service '{}'...".format(service))
    elif remote_port:
        click.echo("Opening tunnel between remote {} port and local {} port"
                   .format(remote_port, local_port if local_port else remote_port))
    dep = seslib.Deployment.load(deployment_id)
    dep.start_port_forwarding(service, node, remote_port, local_port, local_address)


if __name__ == '__main__':
    sesdev_main()
